components/bpmn/org.wso2.carbon.bpmn.analytics.publisher/src/main/java/org/wso2/carbon/bpmn/analytics/publisher/AnalyticsPublisher.java

```diff
@@ -0,0 +1,273 @@
+import org.wso2.carbon.databridge.agent.thrift.exception.AgentException;
+import org.wso2.carbon.databridge.commons.AttributeType;
+import org.wso2.carbon.databridge.commons.StreamDefinition;
+import org.wso2.carbon.databridge.commons.exception.*;
+import org.wso2.carbon.registry.api.Registry;
+import org.wso2.carbon.registry.api.RegistryException;
+import org.wso2.carbon.registry.core.utils.RegistryUtils;
+import org.wso2.carbon.user.api.UserStoreException;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.util.concurrent.Executors;
+
+/**
+ * AnalyticsPublisher uses to publish events to the data receiver in data-bridge
+ */
+public class AnalyticsPublisher {
+	private static Log log = LogFactory.getLog(AnalyticsPublisher.class);
+
+	private String processInstanceStreamId;
+	private String taskInstanceStreamId;
+	private DataPublisher dataPublisher;
+	private AnalyticsPublishServiceUtils analyticsPublishServiceUtils;
+
+	/**
+	 * Initialize the objects for AnalyticsPublisher
+	 */
+	public void initialize() {
+		try {
+			RegistryUtils.setTrustStoreSystemProperties();
+			dataPublisher = createDataPublisher();
+			processInstanceStreamId = getBPMNProcessInstanceStreamId();
+			taskInstanceStreamId = getBPMNTaskInstanceStreamId();
+			analyticsPublishServiceUtils = new AnalyticsPublishServiceUtils();
+			int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+			String tenantDomain = CarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+			Registry registry = BPMNAnalyticsHolder.getInstance().getRegistryService()
+			                                       .getGovernanceSystemRegistry();
+			setPrivilegeContext(tenantId, tenantDomain, registry);
+		} catch (MalformedURLException | AgentException | AuthenticationException | TransportException |
+				DifferentStreamDefinitionAlreadyDefinedException | StreamDefinitionException |
+				MalformedStreamDefinitionException | UserStoreException | RegistryException e) {
+			String errMsg = "Data publisher objects initialization error.";
+			log.error(errMsg, e);
+		}
+	}
+
+	/**
+	 * Set thread local privileges to polling thread
+	 */
+	private void setPrivilegeContext(final int tenantId, final String tenantDomain,
+	                                 final Registry registry) {
+		log.debug("Run setPrivilegeContext method...");
+		Executors.newSingleThreadExecutor().execute(new Runnable() {
+			@Override public void run() {
+				try {
+					PrivilegedCarbonContext.startTenantFlow();
+					PrivilegedCarbonContext privilegedCarbonContext =
+							PrivilegedCarbonContext.getThreadLocalCarbonContext();
+					privilegedCarbonContext.setTenantId(tenantId, true);
+					privilegedCarbonContext.setTenantDomain(tenantDomain, true);
+					privilegedCarbonContext.setRegistry(RegistryType.SYSTEM_GOVERNANCE, registry);
+					doPollingForInstances();
+				} finally {
+					PrivilegedCarbonContext.endTenantFlow();
```

Comment:
I referred the carbon documentation to implement this and documentation has the similar code like what I implemented here. It also not mentioned about a flag to indicate tenant flow started or not. See https://docs.wso2.com/display/Carbon420/CarbonContext+API .